src/System.Collections/src/System/Collections/Generic/BitHelper.cs

```diff
@@ -0,0 +1,153 @@
+    /// samples above; this is the value gotten from ToIntArrayLength(). 
+    /// 
+    /// The length ctor argument is the only exception; for other methods -- MarkBit and 
+    /// IsMarked -- pass in values as indices into the logical bit array, and it will be mapped
+    /// to the position within the array of ints.
+    /// 
+    /// FUTURE OPTIMIZATIONS:
+    /// A method such as FindFirstMarked/Unmarked Bit would be useful for callers that operate 
+    /// on a bit array and then need to loop over it. In particular, if it avoided visiting 
+    /// every bit, it would allow good perf improvements when the bit array is sparse.
+    /// </summary>
+    unsafe internal class BitHelper
+    {   // should not be serialized
+        private const byte MarkedBitFlag = 1;
+        private const byte IntSize = 32;
+
+        // m_length of underlying int array (not logical bit array)
+        private int m_length;
+
+        // ptr to stack alloc'd array of ints
+        [System.Security.SecurityCritical]
+        private int* m_arrayPtr;
+
+        // array of ints
+        private int[] m_array;
+
+        // whether to operate on stack alloc'd or heap alloc'd array 
+        private bool useStackAlloc;
+
+        /// <summary>
+        /// Instantiates a BitHelper with a heap alloc'd array of ints
+        /// </summary>
+        /// <param name="bitArray">int array to hold bits</param>
+        /// <param name="length">length of int array</param>
+        [System.Security.SecurityCritical]
+        internal BitHelper(int* bitArrayPtr, int length)
+        {
+            this.m_arrayPtr = bitArrayPtr;
+            this.m_length = length;
+            useStackAlloc = true;
+        }
+
+        /// <summary>
+        /// Instantiates a BitHelper with a heap alloc'd array of ints
+        /// </summary>
+        /// <param name="bitArray">int array to hold bits</param>
+        /// <param name="length">length of int array</param>
+        internal BitHelper(int[] bitArray, int length)
+        {
+            this.m_array = bitArray;
+            this.m_length = length;
+        }
+
+        /// <summary>
+        /// Mark bit at specified position
+        /// </summary>
+        /// <param name="bitPosition"></param>
+        [System.Security.SecuritySafeCritical]
+        internal unsafe void MarkBit(int bitPosition)
+        {
+            if (useStackAlloc)
+            {
+                int bitArrayIndex = bitPosition / IntSize;
+                if (bitArrayIndex < m_length && bitArrayIndex >= 0)
+                {
+                    m_arrayPtr[bitArrayIndex] |= (MarkedBitFlag << (bitPosition % IntSize));
+                }
+            }
+            else
+            {
+                int bitArrayIndex = bitPosition / IntSize;
+                if (bitArrayIndex < m_length && bitArrayIndex >= 0)
+                {
+                    m_array[bitArrayIndex] |= (MarkedBitFlag << (bitPosition % IntSize));
+                }
```

Comment:
In both this and the subsequent function, the only difference between the if and else block appears to be whether the result is stored into m_arrayPtr or m_array.  You could avoid duplicated code by just having the conditional around that piece rather than the whole thing.